atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/compact/BackgroundCompactor.java

```diff
@@ -0,0 +1,218 @@
+
+    private final CompactionOutcomeMetrics compactionOutcomeMetrics = new CompactionOutcomeMetrics();
+
+    private Thread daemon;
+
+    public static Optional<BackgroundCompactor> createAndRun(TransactionManager transactionManager,
+            KeyValueService keyValueService,
+            RemoteLockService lockService,
+            Supplier<Boolean> inSafeHours) {
+        if (!keyValueService.shouldTriggerCompactions()) {
+            log.info("Not starting a background compactor, because we don't believe our KVS needs one.");
+            return Optional.empty();
+        }
+
+        CompactPriorityCalculator compactPriorityCalculator = CompactPriorityCalculator.create(transactionManager);
+        BackgroundCompactor backgroundCompactor = new BackgroundCompactor(transactionManager,
+                keyValueService,
+                lockService,
+                inSafeHours,
+                compactPriorityCalculator);
+        backgroundCompactor.runInBackground();
+
+        log.info("Started background compactor {}", backgroundCompactor);
+
+        return Optional.of(backgroundCompactor);
+    }
+
+    @VisibleForTesting
+    BackgroundCompactor(TransactionManager transactionManager,
+            KeyValueService keyValueService,
+            RemoteLockService lockService,
+            Supplier<Boolean> inSafeHours,
+            CompactPriorityCalculator compactPriorityCalculator) {
+        this.transactionManager = transactionManager;
+        this.keyValueService = keyValueService;
+        this.lockService = lockService;
+        this.inSafeHours = inSafeHours;
+        this.compactPriorityCalculator = compactPriorityCalculator;
+    }
+
+    @Override
+    public void close() {
+        compactionOutcomeMetrics.registerOccurrenceOf(CompactionOutcome.SHUTDOWN);
+        log.info("Closing BackgroundCompactor");
+        daemon.interrupt();
+        try {
+            daemon.join();
+            daemon = null;
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw Throwables.rewrapAndThrowUncheckedException(e);
+        }
+    }
+
+    private synchronized void runInBackground() {
+        Preconditions.checkState(daemon == null);
+        daemon = new Thread(this::run);
+        daemon.setDaemon(true);
+        daemon.setName("BackgroundCompactor");
+        daemon.start();
+        log.info("Set up the background compactor to be run.");
+    }
+
+    public void run() {
+        log.info("Attempting to start the background compactor for the very first time.");
+        try (SingleLockService compactorLock = createSimpleLocks()) {
+            log.info("Starting background compactor");
+            while (true) {
+                if (Thread.currentThread().isInterrupted()) {
+                    log.warn("Shutting down background compactor because someone interrupted its thread");
+                    Thread.currentThread().interrupt();
```

Comment:
We had a bug in the background sweeper where we weren't spotting InterruptedException in all cases, does a similar bug exist in the background compactor?

Reply:
I've checked all of the try-catch blocks here and it doesn't look like we have a case where we catch an InterruptedException and then don't reset the flag or throw. (Note: in `grabLockAndRunOnce` the last two try-catch blocks don't throw InterruptedException).